pkg/issuer/ca/sign_test.go

```diff
@@ -0,0 +1,147 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ca
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/pem"
+	"testing"
+	"time"
+
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	testpkg "github.com/jetstack/cert-manager/pkg/controller/test"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+	"github.com/jetstack/cert-manager/test/unit/gen"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func generateCSR(t *testing.T, secretKey crypto.Signer) ([]byte, error) {
+	asn1Subj, _ := asn1.Marshal(pkix.Name{
+		CommonName: "test",
+	}.ToRDNSequence())
+	template := x509.CertificateRequest{
+		RawSubject:         asn1Subj,
+		SignatureAlgorithm: x509.SHA256WithRSA,
+	}
+
+	csrBytes, err := x509.CreateCertificateRequest(rand.Reader, &template, secretKey)
+	if err != nil {
+		return nil, err
+	}
+
+	csr := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE REQUEST", Bytes: csrBytes})
+
+	return csr, nil
+}
+
+func generateSelfSignedCertFromCR(t *testing.T, cr *v1alpha1.CertificateRequest, key crypto.Signer,
+	duration time.Duration) (derBytes, pemBytes []byte) {
+	template, err := pki.GenerateTemplateFromCertificateRequest(cr)
+	if err != nil {
+		t.Errorf("error generating template: %v", err)
+	}
+
+	derBytes, err = x509.CreateCertificate(rand.Reader, template, template, key.Public(), key)
+	if err != nil {
+		t.Errorf("error signing cert: %v", err)
+		t.FailNow()
+	}
+
+	pemByteBuffer := bytes.NewBuffer([]byte{})
+	err = pem.Encode(pemByteBuffer, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
+	if err != nil {
+		t.Errorf("failed to encode cert: %v", err)
+		t.FailNow()
+	}
+
+	return derBytes, pemByteBuffer.Bytes()
+}
+
+func TestSign(t *testing.T) {
+	// Build root RSA CA
+	rsaPK := generateRSAPrivateKey(t)
+	rsaPKBytes := pki.EncodePKCS1PrivateKey(rsaPK)
+
+	caCSR, err := generateCSR(t, rsaPK)
+	if err != nil {
+		t.Errorf("failed to generate CA CSR: %s", err)
+		t.FailNow()
+	}
+
+	rootRSACR := gen.CertificateRequest("test-root-ca",
+		gen.SetCertificateRequestCSR(caCSR),
+		gen.SetCertificateRequestIsCA(true),
+		gen.SetCertificateRequestDuration(&metav1.Duration{Duration: time.Hour * 24 * 60}),
+	)
+
+	// generate a self signed root ca valid for 60d
+	_, rsaPEMCert := generateSelfSignedCertFromCR(t, rootRSACR, rsaPK, time.Hour*24*60)
+	rootRSACASecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "root-ca-secret",
+			Namespace: gen.DefaultTestNamespace,
+		},
+		Data: map[string][]byte{
+			corev1.TLSPrivateKeyKey: rsaPKBytes,
+			corev1.TLSCertKey:       rsaPEMCert,
+		},
+	}
+
+	tests := map[string]caFixture{
+		"sign a CertificateRequest": {
```

Comment:
We should increase the test coverage here before we switch the implementation over to use the new Sign method by default. The `Sign` method isn't too complex, so I don't imagine there'll be too many cases to handle currently. I find generating code coverage reports helps to be exhaustive and make sure you cover all different cases.